packages/vm/src/state/transientStorage.ts

```diff
@@ -0,0 +1,131 @@
+import { Address } from 'ethereumjs-util'
+
+export type TStorage = Map<string, Map<string, Buffer>>
+
+export interface TransientStorageModification {
+  addr: Address
+  key: Buffer
+  prevValue: Buffer
+}
+
+export type Changeset = TransientStorageModification[]
+
+export interface TransientStorageOptions {
+  storage?: TStorage
+  changesets?: Changeset[]
+}
+
+export default class TransientStorage {
+  _storage: TStorage
+  _changesets: Changeset[]
+
+  constructor(opts: TransientStorageOptions = {}) {
+    this._storage = opts.storage ?? new Map()
+    this._changesets = opts.changesets ?? [[]]
+  }
+
+  private addModification(modification: TransientStorageModification) {
+    if (this._changesets.length === 0) {
+      throw new Error('no changeset initialized')
+    }
+    const latest = this._changesets[this._changesets.length - 1]
+    latest.push(modification)
+  }
+
+  public get(addr: Address, key: Buffer): Buffer {
+    const map = this._storage.get(addr.toString())
+    if (!map) {
+      return Buffer.alloc(32, 0x00)
+    }
+    const value = map.get(key.toString('hex'))
+    if (!value) {
+      return Buffer.alloc(32, 0x00)
+    }
+    return value
+  }
+
+  public put(addr: Address, key: Buffer, value: Buffer) {
+    if (key.length !== 32) {
+      throw new Error('Transient storage key must be 32 bytes long')
+    }
+
+    if (value.length > 32) {
+      throw new Error('Transient storage value cannot be longer than 32 bytes')
+    }
+
+    if (!this._storage.has(addr.toString())) {
+      this._storage.set(addr.toString(), new Map())
+    }
+    const map = this._storage.get(addr.toString())!
+
+    const str = key.toString('hex')
+    const prevValue = map.get(str) ?? Buffer.alloc(32, 0x00)
+
+    this.addModification({
+      addr,
+      key,
+      prevValue,
+    })
+
+    map.set(str, value)
+  }
+
+  public revert() {
+    const changeset = this._changesets.pop()
+    if (!changeset) {
+      throw new Error('cannot revert without a changeset')
+    }
+    const reversed = changeset.slice().reverse()
```

Comment:
The `slice` copies the array, right? And the reversal seems like it can be optimized, for each key we would only need the first entry in the `changeset`? As that is the original value which we should revert to?

Reply:
According to the docs, `slice` creates a shallow copy
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice
I can refactor this to use a simple for loop that decrements

You need to iterate through the changeset and apply each diff. Perhaps some better naming could help here.

As things are added to transient storage, a `TransientStorageModification` is pushed to the latest `changeset` array. On a `commit`, the latest `changeset` is discarded and on a `revert` the latest `changeset` is used to set each value that was changed to its previous value and then discarded
```js
export type Changeset = TransientStorageModification[]
```

You only need to touch the latest `changeset` and all items in that changeset.

Reply:
the array could be optimized to be a map, since we only care about the first previous value

Reply:
i addressed this in https://github.com/ethereumjs/ethereumjs-monorepo/pull/1768/commits/aa79dcd44a562655398d6180576227b85e338b0f

Reply:
Oh I see what you mean, thanks for the commit @moodysalem